Library/Homebrew/version.rb

```diff
@@ -585,41 +597,63 @@ def <=>(other)
 
     0
   end
-  alias eql? ==
+
+  sig { override.params(other: T.untyped).returns(T::Boolean) }
+  def eql?(other)
+    # Makes sure that the same instance of Version::NULL
+    # will never equal itself; normally Comparable#==
+    # will return true for this regardless of the return
+    # value of #<=>
+    return false if null?
+
+    self == other
+  end
 
   # @api public
   sig { returns(T.nilable(Token)) }
   def major
+    return NULL_TOKEN if null?
+
     tokens.first
   end
 
   # @api public
   sig { returns(T.nilable(Token)) }
   def minor
+    return NULL_TOKEN if null?
+
     tokens.second
   end
 
   # @api public
   sig { returns(T.nilable(Token)) }
   def patch
+    return NULL_TOKEN if null?
+
     tokens.third
   end
 
   # @api public
   sig { returns(T.self_type) }
   def major_minor
-    self.class.new([major, minor].compact.join("."))
+    return self if null?
+
+    major_minor = T.must(tokens[0..1])
+    major_minor.empty? ? NULL : self.class.new(major_minor.join("."))
   end
 
   # @api public
   sig { returns(T.self_type) }
   def major_minor_patch
-    self.class.new([major, minor, patch].compact.join("."))
+    return self if null?
+
+    major_minor_patch = T.must(tokens[0..2])
+    major_minor_patch.empty? ? NULL : self.class.new(major_minor_patch.join("."))
   end
 
   sig { returns(T::Boolean) }
   def empty?
-    version.empty?
+    version&.empty? || false
```

Comment:
```suggestion
    version.blank?
```

Reply:
That's not the same. We want `false` if it is `null?`. If not, we should probably rename this method to `blank?` as well.

Reply:
This method may not actually be used anywhere. The only place where `Version.new` might be called with an empty string is `version ""` in a formula and the emptiness check for that is actually done in a `RuboCop` on a `String`.

Reply:
Cool, that's fine with me, resolving!